--- etc_rc.d_init.d_functions.orig	2008-01-10 23:10:45.000000000 -0600
+++ etc_rc.d_init.d_functions.patched	2008-01-10 23:12:44.000000000 -0600
@@ -94,7 +94,16 @@
 	local remaining sig=
 	local retry=3 count
 
-	remaining=$(LC_ALL=C awk "/^#/ {next} $1" "$2" | sort -r)
+        # Prepare protected filesystems (e.g. device mapper subcomponents of 
+	# rootfs) awk extension source.
+	rootfs_subcomponents_awk_source="{ }"
+	if [ -f /dev/.fstab.live.special ]; then
+	    for protected_fs in $( < /dev/.fstab.live.special ); do
+		rootfs_subcomponents_awk_source="\$2 == \"${protected_fs}\" { next ; } ${rootfs_subcomponents_awk_source}"
+	    done
+	fi
+
+	remaining=$(LC_ALL=C awk --source "${rootfs_subcomponents_awk_source}" --source "/^#/ {next} $1" "$2" | sort -r)
 	while [ -n "$remaining" -a "$retry" -gt 0 ]; do
 		if [ "$retry" -eq 3 ]; then
 			action "$3" fstab-decode umount $5 $remaining
@@ -102,12 +111,12 @@
 			action "$4" fstab-decode umount $5 $remaining
 		fi
 		count=4
-		remaining=$(LC_ALL=C awk "/^#/ {next} $1" "$2" | sort -r)
+		remaining=$(LC_ALL=C awk --source "${rootfs_subcomponents_awk_source}" --source "/^#/ {next} $1" "$2" | sort -r)
 		while [ "$count" -gt 0 ]; do
 			[ -z "$remaining" ] && break
 			count=$(($count-1))
 			usleep 500000
-			remaining=$(LC_ALL=C awk "/^#/ {next} $1" "$2" | sort -r)
+			remaining=$(LC_ALL=C awk --source "${rootfs_subcomponents_awk_source}" --source "/^#/ {next} $1" "$2" | sort -r)
 		done
 		[ -z "$remaining" ] && break
 		fstab-decode /sbin/fuser -k -m $sig $remaining >/dev/null
@@ -122,8 +131,17 @@
 	local remaining devremaining sig=
 	local retry=3
 
-	remaining=$(awk '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $2}' /proc/mounts)
-	devremaining=$(awk '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $1}' /proc/mounts)
+        # Prepare protected filesystems (e.g. device mapper subcomponents of 
+	# rootfs) awk extension source.
+	rootfs_subcomponents_awk_source="{ }"
+	if [ -f /dev/.fstab.live.special ]; then
+	    for protected_fs in $( < /dev/.fstab.live.special ); do
+		rootfs_subcomponents_awk_source="\$2 == \"${protected_fs}\" { next ; } ${rootfs_subcomponents_awk_source}"
+	    done
+	fi
+
+	remaining=$(awk --source "${rootfs_subcomponents_awk_source}" --source '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $2}' /proc/mounts)
+	devremaining=$(awk --source "${rootfs_subcomponents_awk_source}" --source '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $1}' /proc/mounts)
 	while [ -n "$remaining" -a "$retry" -gt 0 ]; do
 		if [ "$retry" -eq 3 ]; then
 			action $"Unmounting loopback filesystems: " \
@@ -137,8 +155,8 @@
 				action $"Detaching loopback device $dev: " \
 				losetup -d $dev
 		done
-		remaining=$(awk '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $2}' /proc/mounts)
-		devremaining=$(awk '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $1}' /proc/mounts)
+		remaining=$(awk --source "${rootfs_subcomponents_awk_source}" --source '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $2}' /proc/mounts)
+		devremaining=$(awk --source "${rootfs_subcomponents_awk_source}" --source '$1 ~ /^\/dev\/loop/ && $2 != "/" {print $1}' /proc/mounts)
 		[ -z "$remaining" ] && break
 		fstab-decode /sbin/fuser -k -m $sig $remaining >/dev/null
 		sleep 3
@@ -511,6 +529,7 @@
 
   STRING=$1
   echo -n "$STRING "
+
   if [ "${RHGB_STARTED:-}" != "" -a -w /etc/rhgb/temp/rhgb-console ]; then
       echo -n "$STRING " > /etc/rhgb/temp/rhgb-console
   fi
