--- usr.share.gdm.gdm-greeter-login-window.glade.orig	2009-07-18 03:03:31.000000000 -0600
+++ usr.share.gdm.gdm-greeter-login-window.glade	2009-07-18 03:04:00.000000000 -0600
@@ -60,7 +60,7 @@
 
 		  <child>
 		    <widget class="GtkImage" id="logo-image">
-		      <property name="visible">True</property>
+		      <property name="visible">False</property>
 		      <property name="icon_name">computer</property>
 		      <property name="pixel_size">64</property>
 		      <property name="xalign">0.5</property>
