--- /tmp/gdm-greeter-login-window.glade	2008-10-29 21:08:55.000000000 -0700
+++ /tmp/new.gdm-greeter-login-window.glade	2008-10-29 21:09:22.000000000 -0700
@@ -17,7 +17,7 @@
             <property name="visible">True</property>
             <child>
               <widget class="GtkImage" id="logo-image">
-                <property name="visible">True</property>
+                <property name="visible">False</property>
                 <property name="pixel_size">64</property>
                 <property name="icon_name">computer</property>
               </widget>
